### DISCUSSION TITLE
How to Trigger a Macro when the printer is powered on?

### DISCUSSION THREAD

**Discussion by sk8board** [mainsail-crew/mainsail#1915] [Help] [answered]
I am using Mainsail and want to start a delayed_gcode macro timer immediately after the printer has been powered on.

[Printer power switch | Mainsail](https://docs.mainsail.xyz/overview/quicktips/printer-power-switch)

https://moonraker.readthedocs.io/en/latest/configuration/#gpio-device-configuration

How do I automatically trigger a macro when the printer is turned on?

**Comment by meteyou** [comment]
you can only execute a "delayed_gcode" when klipper is ready (or at least every macro can only be executed, when klipper is ready): https://www.klipper3d.org/Config_Reference.html#delayed_gcode

**Comment by sk8board** [comment]
Thank you for your reply, as this makes sense. Would it be possible to create a Klipper startup trigger that can be used to trigger a macro immediately after Klipper becomes ready after startup?

**Reply by meteyou** [reply, marked as answer]
yes. pls read the link above. you have only to add `initial_duration` to your delayed_gcode.

**Comment by pmax3** [comment]
Hi, I'd like to pick up this topic as my question is similar. My tasmota power switch works well and as expected. I have that dialog box in mainsail dashboard to power on the printer and switch it off after a print has finished with a delayed macro.
What I miss is a good way to detect, when the socket was switched on externally. There are some udev rules to detect the occurence of the printer and a RESTART is sent to klipper. But this does not work. I belive because of the initial_state: off in the moonraker.conf. Klipper is restarted, but it does not connect to the printer.
When I do not bind the power device to klipper with bound_service: klipper in moonraker.conf the RESTART works and klippers connects to the printer but I don't like not to have the nice button to switch on the printer on the dashboard, when the printer should be powerd from there.
I think it would be nice not to send that RESTART command but a "power on" command. I just don't know how.

Thanks

**Reply by meteyou** [reply]
please always create a new post, unless it's about the exact same problem. your problem actually has very little to do with the current one here and since this post is marked as "solved", your question could go under quite quickly if I overlook the email.

your question has almost nothing to do with moonraker itself, since you only want to operate an external action. as long as moonraker does not receive an active signal from you, it simply does "nothing". so it is only about your udev rule.

please create a new post, upload your current config + udev rule + a Klipper log.

**Reply by pmax3** [reply]
all right, will do so. Thanks
